File: C++/题目1539：师弟.cpp

// 题目1539：师弟

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:56
 * @url:http://ac.jobdu.com/problem.php?pid=1539
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/17581157
 */

#include <stdio.h>
#include <string.h>
#include <queue>

using namespace std;
#define INF 10000000;
const int maxn = 1002;
int path[maxn][maxn];
int minLen[2][maxn];
int visited[maxn];
int n, m, k;

void dijkstra(int s, int rank) {
    int i, j;
    minLen[rank][s] = 0;
    visited[s] = 1;
    int minj = s;
    for (i = 1; i < n + 1; i++) {
        int min = INF;
        for (j = 1; j < n + 1; j++) {
            if (visited[j] == 0 && minLen[rank][j] < min) {
                min = minLen[rank][j];
                minj = j;
            }
        }
        visited[minj] = 1;
        for (j = 1; j < n + 1; j++) {
            if (visited[j] == 0 && minLen[rank][j] > (minLen[rank][minj] + path[minj][j])) {
                minLen[rank][j] = minLen[rank][minj] + path[minj][j];
            }
        }
    }
}

int bfs(int len) {
    queue<int> q;
    while (!q.empty()) q.pop();
    q.push(1);
    int num = 0;
    int allNum = 0;
    visited[1] = 1;
    while (!q.empty()) {
        int size = q.size();//这里是bfs，针对每一个搜索到的点，判断是否为关键路上的点
        if (num > k) {
            return allNum;
        }
        num++;
        while (size--) {
            int s = q.front();
            q.pop();
            if (minLen[0][s] + minLen[1][s] == len) {
                allNum++;
            }
            for (int i = 1; i < n + 1; i++) {
                //	printf("%d\n",path[s][i]);
                if (visited[i] == 0 && path[s][i] < 10000000) {
                    visited[i] = 1;
                    q.push(i);
                }
            }
        }
    }
    return allNum;
}

int main() {
    int i, j;
    while (scanf("%d%d%d", &n, &m, &k) != EOF) {
        for (i = 1; i < n + 1; i++) {
            for (j = 1; j < n + 1; j++) {
                path[i][j] = INF;
            }
        }
        for (i = 0; i < m; i++) {
            int a, b, d;
            scanf("%d%d%d", &a, &b, &d);
            if (path[a][b] > d) {
                path[a][b] = d;
                path[b][a] = d;
            }
        }
        memset(visited, 0, sizeof(visited));
        for (i = 1; i < n + 1; i++) {
            minLen[0][i] = path[1][i];
            minLen[1][i] = path[n][i];
        }
        dijkstra(1, 0);
        memset(visited, 0, sizeof(visited));
        dijkstra(n, 1);
        memset(visited, 0, sizeof(visited));
        int allNum = bfs(minLen[0][n]);
        printf("%d\n", allNum);
    }
    return 0;
}

/**************************************************************
Problem: 1539
User: wangzhenqing
Language: C++
Result: Accepted
Time:650 ms
Memory:4988 kb
****************************************************************/


                        

File: Java/题目1539：师弟.java

// 题目1539：师弟

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:56
 * @url:http://ac.jobdu.com/problem.php?pid=1539
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/17581157
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Main {
    /*
     * 1539
     */
    private static int n, k;

    private static int path[][], visit[];

    public static void main(String[] args) throws Exception {
        StreamTokenizer st = new StreamTokenizer(new BufferedReader(
                new InputStreamReader(System.in)));
        while (st.nextToken() != StreamTokenizer.TT_EOF) {
            n = (int) st.nval;
            st.nextToken();
            int m = (int) st.nval;
            st.nextToken();
            k = (int) st.nval;
            //			初始化路径，都为最大值。
            path = new int[n + 1][n + 1];
            for (int i = 1; i < n + 1; i++) {
                for (int j = 1; j < n + 1; j++) {
                    path[i][j] = Integer.MAX_VALUE;
                }
            }
            for (int i = 0; i < m; i++) {
                st.nextToken();
                int a = (int) st.nval;
                st.nextToken();
                int b = (int) st.nval;
                st.nextToken();
                int d = (int) st.nval;
                if (path[a][b] > d) {
                    path[a][b] = d;
                    path[b][a] = d;
                }
            }

            int minLen1[] = new int[n + 1];
            int minLen2[] = new int[n + 1];
            //			visit初始为0，防止回溯
            visit = new int[n + 1];
            //			初始化1到其他点的距离。
            for (int i = 1; i < n + 1; i++) {
                minLen1[i] = path[1][i];
                minLen2[i] = path[n][i];
            }
            dijkstra(path, minLen1, visit, 1);
            Arrays.fill(visit, 0);
            dijkstra(path, minLen2, visit, n);
            Arrays.fill(visit, 0);
            int allNum = bfs(minLen1, minLen2, minLen1[n]);
            System.out.println(allNum);
        }
    }

private static int bfs(int[] minLen1, int[] minLen2, int len) {
		int allNum = 0;
		Queue<Integer> queue = new LinkedList<Integer>();
        queue.add(1);
        visit[1] = 1;
        int num = 0;
        while (!queue.isEmpty()) {
        	if (num > k) {
        		return allNum; 
			}
        	num++;
        	int size = queue.size();
        	while (size > 0) {
        		int s = queue.poll();
                if (minLen1[s] + minLen2[s] == len) {
                	allNum++;
    			}
                for (int i = 1; i < n+1; i++) {
                	if (visit[i] == 0 && path[s][i] != Integer.MAX_VALUE) {
    					queue.add(i);
    					visit[i] = 1;
    				}
    			}
				size--;
			}
        }
        return allNum;   
	}    private static void dijkstra(int[][] path, int[] minLen, int[] visit, int s) {
        minLen[s] = 0;
        visit[s] = 1;
        for (int i = 1; i < n + 1; i++) {
            int min = Integer.MAX_VALUE;
            int minj = s;
            for (int j = 1; j < n + 1; j++) {
                if (visit[j] == 0 && minLen[j] < min) {
                    min = minLen[j];
                    minj = j;
                }
            }
            visit[minj] = 1;
            for (int j = 1; j < n + 1; j++) {
                if (visit[j] == 0 && minLen[j] > (minLen[minj] + path[minj][j])
                        && minLen[minj] != Integer.MAX_VALUE
                        && path[minj][j] != Integer.MAX_VALUE) {
                    minLen[j] = minLen[minj] + path[minj][j];
                }
            }
        }
    }
}
/**************************************************************
 Problem: 1539
 User: wangzhenqing
 Language: Java
 Result: Accepted
 Time:1980 ms
 Memory:44684 kb
 ****************************************************************/




